app/src/main/kotlin/org/gradle/tools/trace/app/App.kt

```diff
@@ -61,31 +61,58 @@ class ConverterApp : CliktCommand() {
         return File(traceFile.parentFile, traceFile.nameWithoutExtension + format.filePostfix)
     }
 
-    private fun convertToChromeTrace(outputFile: File, slice: BuildOperationTraceSlice) {
-        TraceToChromeTraceConverter().convert(slice, outputFile)
+    private fun convertToChromeTrace(outputFile: File, logs: BuildOperationLogs) {
+        TraceToChromeTraceConverter().convert(logs, outputFile)
     }
 
-    private fun convertToTimelineCsv(outputFile: File, slice: BuildOperationTraceSlice) {
-        TraceToTimelineConverter().convert(slice, outputFile)
+    private fun convertToTimelineCsv(outputFile: File, logs: BuildOperationLogs) {
+        TraceToTimelineConverter().convert(logs, outputFile)
     }
 
-    private fun convertToTransformSummary(outputFile: File, slice: BuildOperationTraceSlice) {
-        TraceToTransformCsvConverter().convert(slice, outputFile)
+    private fun readLogs(traceFile: File): BuildOperationLogs {
+        val logs = readBuildOperationLogs(traceFile)
+        println("Read ${logs.size} build operation logs from ${traceFile.name}")
+        return BuildOperationLogs(logs, include, exclude)
     }
 
-    private fun readTraceSlice(traceFile: File): BuildOperationTraceSlice {
-        val records = readBuildOperationTrace(traceFile)
-        println("Read ${records.size} build operation tree roots from ${traceFile.name}")
-        return BuildOperationTraceSlice(records.toList(), include, exclude)
-    }
-
-    private fun readBuildOperationTrace(traceJsonFile: File): Array<BuildOperationRecord> {
-        val inputTraceJsonText = traceJsonFile.readText()
+    private fun readBuildOperationLogs(traceLogFile: File): List<BuildOperationLog> {
```

Comment:
If the goal of this PR is to avoid reading the file into memory as a string, then we can do that in like two lines without all these changes

```
    private fun readBuildOperationTrace(traceJsonFile: File): Array<BuildOperationRecord> {
        try {
            val reader = Files.newBufferedReader(traceJsonFile.toPath())
            return Gson().fromJson(reader, Array<BuildOperationRecord>::class.java)
        } catch (e: Exception) {
            throw RuntimeException("Failed to read build operation trace file: ${traceJsonFile.absolutePath}", e)
        }
    }
```

I think the real advantage would be to stream the log file line by line and process each individually

Reply:
Now, this would make it more inefficient to create _both_ the summary CSV and the chrome proto file, as we would need to read the JSON twice, but we could also avoid that by visiting all visitors at the same time (process the first event through the first visitor, then the second visitor), and then move on to the next event, etc.

Reply:
I did the change to use a Stream from the file and process both visitors at once.